controller/src/main/java/io/pravega/controller/server/SegmentHelperConnectionManager.java

```diff
@@ -0,0 +1,448 @@
+
+    /**
+     * Shutdown the connection manager where all returned connections are closed and not put back into the
+     * available queue of connections.
+     * It is important to note that even after shutdown is initiated, if `getConnection` is invoked, it will return a connection.
+     */
+    void shutdown() {
+        log.debug("ConnectionManager shutdown initiated");
+
+        // as connections are returned we need to shut them down
+        synchronized (lock) {
+            isRunning = false;
+        }
+        ConnectionObject connection = poll();
+        while (connection != null) {
+            returnConnection(connection);
+            connection = poll();
+        }
+    }
+
+    private ConnectionObject poll() {
+        synchronized (lock) {
+            ConnectionObject polled = availableConnections.poll();
+            while (polled != null) {
+                if (polled.state.get().equals(ConnectionObject.ConnectionState.DISCONNECTED)) {
+                    // discard.. call poll again
+                    polled = availableConnections.poll();
+                } else {
+                    return polled;
+                }
+            }
+            return null;
+        }
+    }
+
+    static class ConnectionObject {
+        private final ClientConnection connection;
+        private final ReusableReplyProcessor processor;
+        private final AtomicReference<ConnectionState> state;
+
+        ConnectionObject(ClientConnection connection, ReusableReplyProcessor processor) {
+            this.connection = connection;
+            this.processor = processor;
+            state = new AtomicReference<>(ConnectionState.CONNECTED);
+        }
+
+        ConnectionState getState() {
+            return state.get();
+        }
+
+        <T> void sendAsync(WireCommand request, CompletableFuture<T> resultFuture) {
+            connection.sendAsync(request, cfe -> {
+                if (cfe != null) {
+                    Throwable cause = Exceptions.unwrap(cfe);
+                    if (cause instanceof ConnectionFailedException) {
+                        resultFuture.completeExceptionally(new WireCommandFailedException(cause, request.getType(),
+                                WireCommandFailedException.Reason.ConnectionFailed));
+                        state.set(ConnectionState.DISCONNECTED);
+                    } else {
+                        resultFuture.completeExceptionally(new RuntimeException(cause));
+                    }
+                }
+            });
+        }
+
+        private enum ConnectionState {
+            CONNECTED,
+            DISCONNECTED
+        }
+    }
+
+    // A reusable reply processor class which can be initialized and uninitialized with new ReplyProcessor. 
+    // This same replyProcessor can be reused with the same connection for handling different replies from servers for
+    // different calls.
+    private static class ReusableReplyProcessor implements ReplyProcessor {
```

Comment:
For now this is fine, but long term should probably be converted to use the RawClient interface. With some review from @shiveshr , @huide9 can make that change as part of a different PR.